torch_agents/cogment_verse_torch_agents/muzero/adapter.py

```diff
@@ -0,0 +1,801 @@
+
+            while not info_queue.empty():
+                try:
+                    info = info_queue.get_nowait()
+                except queue.Empty:
+                    continue
+
+                training_step += 1
+                info["model_version"] = version_info["version_number"]
+                info["training_step"] = training_step
+                info["mean_trial_reward"] = run_total_reward / max(1, trials_completed)
+                info["samples_per_sec"] = total_samples / max(1, time.time() - start_time)
+                info["reanalyzed_samples"] = sum([worker.reanalyzed_samples() for worker in reanalyze_workers])
+                running_stats.update(info)
+
+            if total_samples % config.training.model_publication_interval == 0:
+                version_info = await agent_adapter.publish_version(model_id, agent)
+
+            if total_samples % config.training.log_interval == 0:
+                xp_tracker.log_metrics(
+                    timestamp,
+                    total_samples,
+                    total_samples=total_samples,
+                    **running_stats.get(),
+                )
+                running_stats.reset()
+
+    finally:
+        for worker in workers:
+            worker.terminate()
+
+    log.info(f"[{run_session.params_name}/{run_session.run_id}] finished ({total_samples} samples seen)")
+
+
+class ReplayBufferWorker(mp.Process):
+    def __init__(
+        self,
+        sample_queue,
+        priority_update_queue,
+        batch_queue,
+        reanalyze_queue,
+        reanalyze_update_queue,
+        config,
+        reward_distribution,
+        value_distribution,
+    ):
+        super().__init__()
+        self._sample_queue = sample_queue
+        self._priority_update_queue = priority_update_queue
+        self._batch_queue = batch_queue
+        self._reanalyze_queue = reanalyze_queue
+        self._reanalyze_update_queue = reanalyze_update_queue
+        self._replay_buffer_size = mp.Value(ctypes.c_uint32, 0)
+        self._training_config = config
+        self._device = config.train_device
+        self.reward_distribution = reward_distribution
+        self.value_distribution = value_distribution
+
+    def run(self):
+        episode_samples = {}
+        replay_buffer = TrialReplayBuffer(
+            max_size=self._training_config.max_replay_buffer_size,
+            discount_rate=self._training_config.discount_rate,
+            bootstrap_steps=self._training_config.bootstrap_steps,
+        )
+
+        zero_reward_probs = self.reward_distribution.compute_target(torch.tensor(0.0)).cpu().detach()
+        zero_value_probs = self.value_distribution.compute_target(torch.tensor(0.0)).cpu().detach()
+
+        while True:
+            # Fetch & perform all pending priority updates
+            while not self._priority_update_queue.empty():
+                try:
+                    episodes, steps, priorities = self._priority_update_queue.get_nowait()
```

Comment:
You are relying a lot on get_nowait() because your workers have several queues as input. because of `get_nowait` you are doing a lot of, basically polling, instead of relying on the "natural" orchestration that would be given by the queues Maybe there's a way to basically have simpler workers that acts as consumer of a single queue.